90zfsbootmenu/zfsbootmenu-lib.sh

```diff
@@ -977,10 +977,19 @@ populate_be_list() {
       # All other datasets are ignored
       continue
     fi
+    if [ -n "${BOOTFS}" ] && [ "${BOOTFS}" = "${fs}" ] ; then
+      # If BOOTFS is defined, we'll manually append it to the array
+      continue
+    fi
+
     candidates+=( "${fs}" )
   done <<< "$(zfs list -H -o name,mountpoint,org.zfsbootmenu:active)"
 
-  for fs in "${candidates[@]}"; do
+  # Reverse sort the list, to line up with our global usage of --tac
+  readarray -t sorted < <( for env in "${candidates[@]}" ; do echo "$env" ; done | sort -r )
+  [ -n "${BOOTFS}" ] && sorted+=( "${BOOTFS}" )
```

Comment:
If the `bootfs` property is set on a pool but the target filesystem doesn't exist, I believe we will still have a nonempty `${BOOTFS}`. This means we might potentially add a BE to the list that doesn't exist. If we add `have_bootfs` as suggested in Line 981, the test becomes `[ -n "${have_bootfs}" ]` to ensure we only add a BE that exists.

Reply:
In my testing, `bootfs` always has to exist, if it is defined. `zpool` won't let me assign a value to `bootfs` if the filesystem doesn't exist. If the `bootfs` filesystem is deleted, the pool property is automatically cleared.

If `bootfs` points to a filesystem that is not a valid BE, the tests that process the candidates array will catch it and discard it.

Reply:
You're right, there is no value in omitting a non-empty `$BOOTFS` since it will be validated by `find_be_kernels` later in this function.